syz-verifier/exectask.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/syzkaller/pkg/rpctype"
+	"github.com/google/syzkaller/prog"
+)
+
+type EnvDescr int64
+
+const (
+	AnyEnvironment EnvDescr = iota
+	NewEnvironment
+	// TODO: add CleanVMEnvironment support.
+
+	EnvironmentsCount
+)
+
+// ExecTask is the atomic analysis entity. Once executed, it could trigger the
+// pipeline propagation fof the program.
+type ExecTask struct {
+	CreationTime   time.Time
+	Program        *prog.Prog
+	ID             int64
+	ExecResultChan ExecResultChan
+
+	priority int // The priority of the item in the queue.
+	// The index is needed by update and is maintained by the heap.Interface methods.
+	index int // The index of the item in the heap.
+}
+
+func (t *ExecTask) ToRPC() *rpctype.ExecTask {
+	return &rpctype.ExecTask{
+		Prog: t.Program.Serialize(),
+		ID:   t.ID,
+	}
+}
+
+var (
+	ChanMapMutex           = sync.Mutex{}
+	TaskIDToExecResultChan = map[int64]ExecResultChan{}
+	TaskCounter            = int64(-1)
+)
+
+type ExecResultChan chan *ExecResult
+
+func MakeExecTask(prog *prog.Prog) *ExecTask {
+	task := &ExecTask{
+		CreationTime:   time.Now(),
+		Program:        prog,
+		ExecResultChan: make(ExecResultChan),
+		ID:             atomic.AddInt64(&TaskCounter, 1),
+	}
+
+	ChanMapMutex.Lock()
+	defer ChanMapMutex.Unlock()
+	TaskIDToExecResultChan[task.ID] = task.ExecResultChan
+
+	return task
+}
+
+func DeleteExecTask(task *ExecTask) {
+	ChanMapMutex.Lock()
+	defer ChanMapMutex.Unlock()
+	delete(TaskIDToExecResultChan, task.ID)
+}
+
+func GetExecResultChan(taskID int64) ExecResultChan {
+	ChanMapMutex.Lock()
+	defer ChanMapMutex.Unlock()
+
+	return TaskIDToExecResultChan[taskID]
```

Comment:
This should also remove the task from TaskIDToExecResultChan.
This will be less code, simpler and less error prone.
Currently it's possible to find the task in TaskIDToExecResultChan several times, but only for a short, non-deterministic period of time (while it's not removed by somebody else out there asynchronously). We don't really want to find/send twice, right?

Reply:
Hm... Still believe it is better to keep "delete" where it is now.
Calling side is in charge of the task lifetime. 
Reading the calling side code it is clear when and how object is deleted.

Reply:
Just don't see the different way to implement it. Lets keep as is.